### PR TITLE
MAINT Factor GraphBuilder class out of the build_from_graph function

### PR DESCRIPTION
**Pull request by hoodmane**
This rewrites all of the closed-over state in build_from_graph into a separate GraphBuilder class. Should make it easier to understand the code here.

### CODE REVIEW

File: pyodide-build/pyodide_build/buildall.py [hunk 1]
```diff
@@ -497,6 +498,138 @@ def generate_needs_build_set(
     return needs_build
 
 
+@dataclasses.dataclass(eq=False, repr=False)
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:501]: Why is `dataclass` used? It is not as a container for the data.

File: pyodide-build/pyodide_build/buildall.py [hunk 2]
```diff
@@ -497,6 +498,138 @@ def generate_needs_build_set(
...
+        while True:
+            try:
+                return self.build_queue.get(timeout=0.5)[1]
+            except Empty:
+                pass
+
+    def builder(self, n: int) -> None:
+        while True:
+            pkg = self.get_pkg()
+            print(n, "starting", pkg.name)
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:591]: I guess this was for debugging?

File: pyodide-build/pyodide_build/buildall.py [hunk 3]
```diff
@@ -497,6 +498,138 @@ def generate_needs_build_set(
...
+            with self.queue_index(pkg) as idx:
+                if idx is None:
+                    # Rust package and we're already building one.
+                    # Release the GIL so new packages get queued
+                    sleep(0.01)
+                    continue
+                pkg._queue_idx = idx
+                res = self.build_one(n, pkg)
+                print(n, "finished", pkg.name, not bool(res))
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:600]: ditto

File: pyodide-build/pyodide_build/buildall.py [hunk 4]
```diff
@@ -497,6 +498,138 @@ def generate_needs_build_set(
...
+        try:
+            yield
+        except BaseException:
+            success = False
+            raise
+        finally:
+            pkg_status.finish(success, perf_counter() - t0)
+            self.progress_formatter.remove_package(pkg_status)
+
+    def build_one(self, n: int, pkg: BasePackage) -> BaseException | None:
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:571]: Could you please add underscore to non-public methods?

**Review by ryanking13 — COMMENTED** [review]
Thanks for the maintenance work!

File: pyodide-build/pyodide_build/buildall.py (hunk 1 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:501]: It was originally. But yeah I'll remove it.

File: pyodide-build/pyodide_build/buildall.py [hunk 5]
```diff
@@ -496,6 +497,163 @@ def generate_needs_build_set(
...
+    pkg_map: dict[str, BasePackage]
+    build_args: BuildArgs
+    build_dir: Path
+    needs_build: set[str]
+    build_queue: PriorityQueue[tuple[int, BasePackage]]
+    built_queue: Queue[tuple[BasePackage, BaseException | None]]
+    lock: Lock
+    building_rust_pkg: bool
+    queue_idx: int
+    progress_formatter: ReplProgressFormatter
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:519]: As it is not a dataclass anymore, I guess these variables don't need to be defined in a class variable.

**Review by ryanking13 — APPROVED** [review]
Thanks!

File: pyodide-build/pyodide_build/buildall.py (hunk 5 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:519]: Isn't this just the normal way to do type annotations for class attributes?

Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:519]: I guess not. Thanks.

**Pull request merged**